File: sdk/server/node/src/sse.ts
/**
 * `GET /v1/events` server-sent events as an async iterator, with automatic reconnection and
 * `Last-Event-ID` resume.
 */

import { AurixError, AurixNetworkError } from "./errors.js";
import type { EventEnvelope } from "./generated/types.js";
import type { AurixHttp, RequestOptions } from "./http.js";

export interface SseEvent {
  /** `event:` field — an event type, `stream.open`, or `lagged`. */
  type: string;
  /** `id:` field (event id) when present. */
  id: string | undefined;
  /** Parsed `data:` (JSON) — an `EventEnvelope` for regular events; raw string when not JSON. */
  data: EventEnvelope | unknown;
  raw: string;
}

export interface EventStreamOptions extends RequestOptions {
  /** Comma-separated (or array of) event types; default: everything except high-frequency ones. */
  types?: string | string[];
  /** Reconnect on stream end / network error (default true). Non-retryable HTTP errors always throw. */
  reconnect?: boolean;
  /** Initial reconnect delay in ms (default 1000, doubles up to `maxReconnectDelayMs`). */
  reconnectDelayMs?: number;
  maxReconnectDelayMs?: number;
  /** Resume point sent as `Last-Event-ID` on the first connection. */
  lastEventId?: string;
}

/**
 * Iterates the application's event stream. Stop by aborting `options.signal` or breaking out of
 * the `for await` loop. On `lagged` fetch `/v1/events/snapshot` to resynchronise.
 */
export async function* eventStream(client: AurixHttp, options: EventStreamOptions = {}): AsyncGenerator<SseEvent, void, undefined> {
  const fetchImpl = client.fetchImpl;
  const types = Array.isArray(options.types) ? options.types.join(",") : options.types;
  const url = client.url("/v1/events", types ? { types } : undefined);
  let lastEventId = options.lastEventId;
  let delay = options.reconnectDelayMs ?? 1000;
  const maxDelay = options.maxReconnectDelayMs ?? 30_000;
  const reconnect = options.reconnect ?? true;
  for (;;) {
    if (options.signal?.aborted) return;
    const headers: Record<string, string> = {
      Accept: "text/event-stream",
      ...client.opts.headers,
      ...client.authHeaders(options.auth ?? client.opts),
      ...options.headers,
    };
    if (lastEventId) headers["Last-Event-ID"] = lastEventId;
    let res: Response;
    try {
      const init: RequestInit = { method: "GET", headers };
      if (options.signal) init.signal = options.signal;
      res = await fetchImpl(url, init);
    } catch (err) {
      if (options.signal?.aborted) return;
      if (!reconnect) throw new AurixNetworkError(err instanceof Error ? err.message : String(err), "GET", "/v1/events", err);
      await sleep(delay, options.signal);
      delay = Math.min(maxDelay, delay * 2);
      continue;
    }
    if (!res.ok) {
      const text = await res.text();
      const error = AurixError.fromResponse(res.status, res.headers.get("content-type") ?? "", text, res.headers, "GET", "/v1/events");
      if (reconnect && (res.status === 429 || res.status >= 500)) {
        await sleep(error.retryAfterMs ?? delay, options.signal);
        delay = Math.min(maxDelay, delay * 2);
        continue;
      }
      throw error;
    }
    if (!res.body) throw new AurixNetworkError("event stream has no body", "GET", "/v1/events");
    delay = options.reconnectDelayMs ?? 1000;
    try {
      for await (const ev of parseSse(res.body, options.signal)) {
        if (ev.id !== undefined) lastEventId = ev.id;
        if (ev.retry !== undefined) delay = ev.retry;
        if (ev.type === undefined && ev.data === "") continue;
        let data: unknown = ev.data;
        try {
          data = JSON.parse(ev.data);
        } catch {
          /* keep raw */
        }
        yield { type: ev.type ?? "message", id: ev.id, data, raw: ev.data };
      }
    } catch (err) {
      if (options.signal?.aborted) return;
      if (!reconnect) throw err;
    }
    if (options.signal?.aborted || !reconnect) return;
    await sleep(delay, options.signal);
    delay = Math.min(maxDelay, delay * 2);
  }
}

interface RawSse {
  type: string | undefined;
  id: string | undefined;
  data: string;
  retry: number | undefined;
}

/** Minimal SSE parser (comments, multi-line `data:`, `id:`, `event:`, `retry:`). */
export async function* parseSse(body: ReadableStream<Uint8Array>, signal?: AbortSignal): AsyncGenerator<RawSse, void, undefined> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let cur: RawSse = { type: undefined, id: undefined, data: "", retry: undefined };
  let dataLines: string[] = [];
  const onAbort = () => void reader.cancel().catch(() => undefined);
  signal?.addEventListener("abort", onAbort, { once: true });
  try {
    for (;;) {
      const { value, done } = await reader.read();
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buffer.search(/\r\n|\n|\r/)) >= 0) {
        const line = buffer.slice(0, nl);
        buffer = buffer.slice(nl + (buffer.startsWith("\r\n", nl) ? 2 : 1));
        if (line === "") {
          if (dataLines.length > 0 || cur.type !== undefined || cur.id !== undefined) {
            cur.data = dataLines.join("\n");
            yield cur;
          }
          cur = { type: undefined, id: undefined, data: "", retry: undefined };
          dataLines = [];
          continue;
        }
        if (line.startsWith(":")) continue;
        const colon = line.indexOf(":");
        const field = colon < 0 ? line : line.slice(0, colon);
        let value = colon < 0 ? "" : line.slice(colon + 1);
        if (value.startsWith(" ")) value = value.slice(1);
        switch (field) {
          case "event":
            cur.type = value;
            break;
          case "data":
            dataLines.push(value);
            break;
          case "id":
            if (!value.includes("\0")) cur.id = value;
            break;
          case "retry": {
            const n = Number(value);
            if (Number.isInteger(n) && n >= 0) cur.retry = n;
            break;
          }
          default:
            break;
        }
      }
    }
  } finally {
    signal?.removeEventListener("abort", onAbort);
    reader.releaseLock();
  }
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve) => {
    if (signal?.aborted) return resolve();
    const t = setTimeout(() => {
      signal?.removeEventListener("abort", done);
      resolve();
    }, ms);
    const done = () => {
      clearTimeout(t);
      resolve();
    };
    signal?.addEventListener("abort", done, { once: true });
  });
}

File: sdk/server/go/sse.go
package aurix

import (
	"bufio"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

// SSEEvent is one message from `GET /v1/events`.
type SSEEvent struct {
	// Type is the `event:` field — an event type, "stream.open" or "lagged".
	Type string
	// ID is the `id:` field when present.
	ID string
	// Data is the raw `data:` payload; for regular events it is an EventEnvelope (see Envelope).
	Data []byte
}

// Envelope decodes Data as an EventEnvelope.
func (e SSEEvent) Envelope() (*EventEnvelope, error) {
	var env EventEnvelope
	if err := json.Unmarshal(e.Data, &env); err != nil {
		return nil, err
	}
	return &env, nil
}

// EventStreamOptions configures Client.Events.
type EventStreamOptions struct {
	// Types filters event types (server default: everything except high-frequency events).
	Types []string
	// Reconnect re-establishes the stream after it ends or fails (default true; set NoReconnect).
	NoReconnect bool
	// ReconnectDelay is the initial delay (default 1 s, doubling up to MaxReconnectDelay).
	ReconnectDelay    time.Duration
	MaxReconnectDelay time.Duration
	// LastEventID resumes from a known position on the first connection.
	LastEventID string
	// Request options (headers, credentials).
	RequestOptions []RequestOption
}

// Events streams the application's events, calling fn for each one until ctx is cancelled, fn
// returns an error, or a non-retryable HTTP error occurs. On a "lagged" event fetch
// `/v1/events/snapshot` to resynchronise.
func (c *Client) Events(ctx context.Context, opts EventStreamOptions, fn func(SSEEvent) error) error {
	q := url.Values{}
	if len(opts.Types) > 0 {
		q.Set("types", strings.Join(opts.Types, ","))
	}
	u := c.URL("/v1/events", q)
	delay := opts.ReconnectDelay
	if delay <= 0 {
		delay = time.Second
	}
	maxDelay := opts.MaxReconnectDelay
	if maxDelay <= 0 {
		maxDelay = 30 * time.Second
	}
	initial := delay
	lastID := opts.LastEventID
	var rc requestConfig
	for _, o := range opts.RequestOptions {
		o(&rc)
	}
	// The stream is long-lived: never apply the JSON client's overall timeout to it.
	httpClient := &http.Client{Transport: c.http.Transport, CheckRedirect: c.http.CheckRedirect, Jar: c.http.Jar}
	for {
		if ctx.Err() != nil {
			return ctx.Err()
		}
		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
		if err != nil {
			return err
		}
		req.Header.Set("Accept", "text/event-stream")
		req.Header.Set("User-Agent", c.userAgent)
		for k, vs := range c.headers {
			for _, v := range vs {
				req.Header.Add(k, v)
			}
		}
		if rc.creds != nil {
			rc.creds.apply(req.Header)
		} else {
			c.creds.apply(req.Header)
		}
		for k, vs := range rc.headers {
			for _, v := range vs {
				req.Header.Add(k, v)
			}
		}
		if lastID != "" {
			req.Header.Set("Last-Event-ID", lastID)
		}
		resp, err := httpClient.Do(req)
		if err != nil {
			if ctx.Err() != nil {
				return ctx.Err()
			}
			if opts.NoReconnect {
				return &NetworkError{Method: "GET", Path: "/v1/events", Err: err}
			}
			if err := sleepCtx(ctx, delay); err != nil {
				return err
			}
			delay = min(maxDelay, delay*2)
			continue
		}
		if resp.StatusCode < 200 || resp.StatusCode >= 300 {
			body, _ := io.ReadAll(io.LimitReader(resp.Body, 1<<16))
			_ = resp.Body.Close()
			ct := resp.Header.Get("Content-Type")
			apiErr := errorFromResponse(resp.StatusCode, ct, body, resp.Header, "GET", "/v1/events")
			if !opts.NoReconnect && (resp.StatusCode == http.StatusTooManyRequests || resp.StatusCode >= 500) {
				d := delay
				if apiErr.RetryAfter > 0 {
					d = apiErr.RetryAfter
				}
				if err := sleepCtx(ctx, d); err != nil {
					return err
				}
				delay = min(maxDelay, delay*2)
				continue
			}
			return apiErr
		}
		delay = initial
		err = readSSE(resp.Body, func(ev sseRaw) error {
			if ev.id != "" {
				lastID = ev.id
			}
			if ev.retry > 0 {
				delay = time.Duration(ev.retry) * time.Millisecond
			}
			typ := ev.typ
			if typ == "" {
				typ = "message"
			}
			return fn(SSEEvent{Type: typ, ID: ev.id, Data: []byte(ev.data)})
		})
		_ = resp.Body.Close()
		if err != nil && ctx.Err() == nil && !isStreamEnd(err) {
			return err
		}
		if ctx.Err() != nil {
			return ctx.Err()
		}
		if opts.NoReconnect {
			return nil
		}
		if err := sleepCtx(ctx, delay); err != nil {
			return err
		}
		delay = min(maxDelay, delay*2)
	}
}

type sseRaw struct {
	typ   string
	id    string
	data  string
	retry int
}

type streamEndError struct{ err error }

func (e streamEndError) Error() string { return e.err.Error() }

func isStreamEnd(err error) bool {
	_, ok := err.(streamEndError)
	return ok
}

// readSSE parses text/event-stream from r (comments, multi-line data, id, event, retry) and
// calls emit per event. Transport errors mid-stream are wrapped as streamEndError so the
// caller can reconnect; errors from emit are returned as-is.
func readSSE(r io.Reader, emit func(sseRaw) error) error {
	sc := bufio.NewScanner(r)
	sc.Buffer(make([]byte, 0, 64*1024), 4*1024*1024)
	var cur sseRaw
	var data []string
	dispatch := func() error {
		if len(data) == 0 && cur.typ == "" && cur.id == "" {
			cur, data = sseRaw{}, nil
			return nil
		}
		cur.data = strings.Join(data, "\n")
		err := emit(cur)
		cur, data = sseRaw{}, nil
		return err
	}
	for sc.Scan() {
		line := strings.TrimRight(sc.Text(), "\r")
		if line == "" {
			if err := dispatch(); err != nil {
				return err
			}
			continue
		}
		if strings.HasPrefix(line, ":") {
			continue
		}
		field, value, _ := strings.Cut(line, ":")
		value = strings.TrimPrefix(value, " ")
		switch field {
		case "event":
			cur.typ = value
		case "data":
			data = append(data, value)
		case "id":
			if !strings.ContainsRune(value, 0) {
				cur.id = value
			}
		case "retry":
			if n, err := strconv.Atoi(value); err == nil && n >= 0 {
				cur.retry = n
			}
		}
	}
	if err := sc.Err(); err != nil {
		return streamEndError{err}
	}
	return nil
}
